docs/book/src/tasks/experimental-features/runtime-sdk/implement-topology-mutation-hook.md

```diff
@@ -35,11 +38,116 @@ External patches have the following advantages:
+    # if metadata.generation isn't the same as observedGeneration Cluster using the ClusterClass should not reconcile.
+    observedGeneration: xx
+    # variables contains a list of all variable definitions, both inline and from external patches, that belong to the ClusterClass.
+    variables:
+      - name: no-proxy
+        definitions:
+          - namespace: inline
+            required: true
+            schema:
+              openAPIV3Schema:
+                type: string
+                default: "internal.com"
+                example: "internal.com"
+                description: "comma-separated list of machine or domain names excluded from using the proxy."
+      - name: http-proxy
+        definitions:
+          - namespace: inline
+            schema:
+              openAPIV3Schema:
+                type: string
+                default: "proxy.example.com"
+                example: "proxy.example.com"
+                description: "proxy for http calls."
+          - namespace: lbImageName
+            schema:
+              openAPIV3Schema:
+                type: string
+                default: "different.example.com"
+                example: "different.example.com"
+                description: "proxy for http calls."
+```
+
+### Variable namespacing
+Variable definitions can be inline in the ClusterClass or from any number of external DiscoverVariables hooks. The source 
+of a variable definition is recorded in the `namespaces` field in ClusterClass `.status.variables`.
+Variables that are defined by an external DiscoverVariables hook will have the name of the patch they are associated with as their namespace.
+Variables that are defined in the ClusterClass `.spec.variables` will have the namespace `inline`.
+Note: `inline` is a reserved namespace. It can not be used as the name of an external patch to avoid conflicts.
+
+If all variables that share a name have equivalent schemas the variables are considered `global` . `global` variables can
+be set without providing a namespace - [see below](#setting-values-for-variables-in-the-cluster). The CAPI components will
+consider variable definitions to be equivalent when they share a name and their schema is exactly equal.
+
+### Setting values for variables in the Cluster
+Setting variables that are defined with external variable definitions requires attention to be paid to variable namespacing, as exposed in the ClusterClass status. 
+Variable values are set in Cluster `.spec.topology.variables`.
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: Cluster
+#metadata 
+spec:
+    topology:
+        variables:
+        - name: no-proxy
+          value: "internal.domain.com"
+          # namespace is not needed as this variable is in the global namespace.
+        - name: http-proxy
+          namespace: inline
+          value: "proxy.example2.com"
+        - name: http-proxy
+          namespace: lbImageRepository
+          value: "proxy.example2.com"
+```
```

Comment:
what about
```suggestion
Variable definitions can be inline in the ClusterClass or the result of calls to  DiscoverVariables hooks. 
The system provides support for handling the same variable being defined in multiple places, and this support is designed around two principles:
- Keeping user UX as simple as possible.  
   The user just wants to provide the value for each variables once, they should not be bothered by the fact that this same variable is defined more than once or used in different ways.
- ClusterClass authors and Runtime Extension authors are responsible for acting proactively in order to ensure the best UX for the users. If they fail in doing so the user UX will degrade accordingly

I practice this means:
- All variables definitions will be assigned to a namespace:
   - Variable definitions defined in the ClusterClass are assigned to the inline namespace.  
   - Variable definitions resulting from external DiscoverVariable hooks are associated with a namespace having the same name as the patch defining the hook call
   - NOTE: `.status.variables` field in ClusterClass contains the result of this process
   - NOTE: `inline` becomes a reserved word that cannot be used anymore as a patch name.
   
 - If all variables definitions that share a name have equivalent schemas the definitions are considered `global`
    - Users can provide a single value (without namespace) and this will be used for all the corresponding definitions
    - NOTE: Variable definitions are equivalent when they share a name and their schema is exactly equal.  
 - If variables that share a name do not have equivalent schemas they are not `global`. 
    - Users must provide a single value for each variable definition, identified by the namespace (degraded UX).
    

Below is an example of how to set global and namespaced variables in a Cluster:

Setting variables using external variables requires attention to be paid to variable namespacing, as exposed in the ClusterClass status. 
Variable values are set in Cluster `.spec.topology.variables`.

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: Cluster
#metadata 
spec:
    topology:
        variables:
        # namespace is not needed as this variable is global; this also allows migration of existing variables.
        - name: no-proxy
          value: "internal.domain.com"
        # namespaced variables (the value must be provided twice according the two different schemas.
        - name: http-proxy
          namespaces: patch1
          value: http://proxy.example2.com:1234
       - name: http-proxy
          namespaces: patch2
          value:
             host: proxy.example2.com
             port: 1234
```

```

so we document the why we are doing all this work; also changed the example a little bit, PTAL

Reply:
IMO there's too much detail here around the reasoning for this - frontrunning potential comments. I think it's better to just state what users should do, and what situations it covers instead of principles in this doc - given it's for implementers.

If you think this is really useful though maybe it would fit better in the proposal?

Reply:
ACK, and I don't want to block it for the first iteration
just please take a look at the example, if it makes sense to use it (it makes it more explicit that schemas are not equal)

Reply:
I did take the example :smile: - it really does make it more clear.